File: 2022/05/part2.rb
#!/usr/bin/env ruby

input = ARGV[0] == "ex" ? "example" : "input"

it = IO.foreach(File.join(__dir__, input), chomp: true)
game = it.take_while { !_1.empty? }
instructions = it.drop_while { !_1.empty? }.drop(1)

# Parse Game

size = game.last.split.size
game = game[0...-1].map { |line|
  size.times.map { |i| line[i * 4 + 1]  } # 1 5 9
}.reverse.transpose.map { |col| col.reject { _1 == " " } }

# Parse Instructions

# intructions are [num, from, to] arrays of one move.
instructions = instructions.map { |l|l.scan(/\d+/).map(&:to_i).then {[_1, _2 - 1, _3 - 1]} }

# Play Game

instructions.each do
  game[_3].push *game[_2].pop(_1)
end

puts game.map(&:last).join

File: 2024/01/part2.py
#!/usr/bin/env python3

import fileinput

my_list: list[list[int]] = list()

for line in fileinput.input():
    my_list.append(list(map(lambda x: int(x), line.split())))


group_a: list[int] = list()
group_b: list[int] = list()

group_a = [a for (a, _) in my_list]
group_b = [b for (_, b) in my_list]


# print(group_a)
# print(group_b)

sum = 0

internal_sums: dict[int, int] = {}

for i in range(len(group_a)):
    num = group_a[i]
    if num not in internal_sums:
        internal_sums[num] = group_b.count(num)
    sum += num * internal_sums[num]

print(sum)
